tests/capture/captest.c

```diff
@@ -0,0 +1,239 @@
+			    got_captures[i].pos[0]);
+			FAIL("capture mismatch");
+		}
+		if (got_captures[i].pos[1] != exp_captures[i].pos[1]) {
+			fprintf(stderr, "capture[%lu].pos[1]: exp %lu, got %lu\n",
+			    i, exp_captures[i].pos[1],
+			    got_captures[i].pos[1]);
+			FAIL("capture mismatch");
+		}
+	}
+
+	fsm_free(fsm);
+	return 0;
+}
+
+struct fsm *
+captest_fsm_of_string(const char *string, unsigned end_id)
+{
+	struct fsm *fsm = captest_fsm_with_options();
+	const size_t length = strlen(string);
+	size_t i;
+	struct captest_end_opaque *eo = NULL;
+
+	if (fsm == NULL) {
+		return NULL;
+	}
+
+	eo = calloc(1, sizeof(*eo));
+	if (eo == NULL) {
+		goto cleanup;
+	}
+
+	eo->tag = CAPTEST_END_OPAQUE_TAG;
+
+	/* set bit for end state */
+	assert(end_id < 8*sizeof(eo->ends));
+	eo->ends |= (1U << end_id);
+
+	if (!fsm_addstate_bulk(fsm, length + 1)) {
+		goto cleanup;
+	}
+	fsm_setstart(fsm, 0);
+
+	for (i = 0; i < length; i++) {
+		if (!fsm_addedge_literal(fsm, i, i + 1, string[i])) {
+			goto cleanup;
+		}
+	}
+	fsm_setend(fsm, length, 1);
+	fsm_setopaque(fsm, length, eo);
+
+	return fsm;
+
+cleanup:
+	if (eo != NULL) { free(eo); }
+	fsm_free(fsm);
+	return NULL;
+}
+
+static struct fsm_options options;
+
+static void captest_carryopaque(struct fsm *src_fsm,
+    const fsm_state_t *src_set, size_t n,
+    struct fsm *dst_fsm, fsm_state_t dst_state)
+{
+	struct captest_end_opaque *eo_src = NULL;
+	struct captest_end_opaque *eo_dst = NULL;
+	struct captest_end_opaque *eo_old_dst = NULL;
+	size_t i;
+	const int log_level = 0;
+
+	if (log_level > 0) {
+		fprintf(stderr, "captest_carryopaque: src_fsm %p, src_set %p, n %lu, dst_fsm %p, dst_state %u\n",
+		    (void *)src_fsm, (void *)src_set, n,
+		    (void *)dst_fsm, dst_state);
+	}
+
+	eo_old_dst = fsm_getopaque(dst_fsm, dst_state);
+
+	eo_dst = calloc(1, sizeof(*eo_dst));
+	/* FIXME: no way to handle an alloc error in carryopaque? */
+	assert(eo_dst != NULL);
+	eo_dst->tag = CAPTEST_END_OPAQUE_TAG;
+	if (log_level > 0) {
+		fprintf(stderr, "captest_carryopaque: new opaque %p (eo_dst)\n",
+		    (void *)eo_dst);
+	}
+
+	if (eo_old_dst != NULL) {
+		assert(eo_old_dst->tag == CAPTEST_END_OPAQUE_TAG);
+		eo_dst->ends |= eo_old_dst->ends;
+		if (log_level > 0) {
+			fprintf(stderr, "carryopaque: old_dst (%d) ends 0x%x\n",
+			    dst_state, eo_dst->ends);
+		}
+	}
+
+	fsm_setopaque(dst_fsm, dst_state, eo_dst);
+
+	/* union bits set in eo_src->ends into eo_dst->ends */
+	for (i = 0; i < n; i++) {
+		eo_src = fsm_getopaque(src_fsm, src_set[i]);
+		if (eo_src == NULL) {
+			continue;
+		}
+		if (log_level > 0) {
+			fprintf(stderr, "carryopaque: dst %p (%d) <- src[%lu] (%d) %p\n",
+			    (void *)eo_dst, dst_state, i, src_set[i],
+			    (void *)eo_src);
+		}
+		assert(eo_src->tag == CAPTEST_END_OPAQUE_TAG);
+		eo_dst->ends |= eo_src->ends;
+		if (log_level > 0) {
+			fprintf(stderr, "carryopaque: eo_src ends -> 0x%x\n",
+			    eo_dst->ends);
+		}
+
+		fsm_setopaque(src_fsm, src_set[i], NULL);
```

Comment:
After some further work I'm not sure this actually should be clearing the src state's opaque here. Removing this seems sufficient to fix the suppressed failure in `capture4.c`, because it's no longer clearing the opaque and the `abc` end state is now considered a match for both `abc` and `ab*c`.

I'm not 100% sure this is correct, because I don't fully understand the invariants behind carryopaque yet. Does that make sense?